File: src/main/java/io/github/chaosdave34/kitpvp/enchantments/LifeStealEnchantment.java
package io.github.chaosdave34.kitpvp.enchantments;

import io.github.chaosdave34.ghutils.enchantment.CustomEnchantment;
import net.minecraft.tags.ItemTags;
import net.minecraft.world.entity.EquipmentSlot;
import org.bukkit.attribute.Attribute;
import org.bukkit.attribute.AttributeInstance;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.inventory.ItemStack;

public class LifeStealEnchantment extends CustomEnchantment {
    public LifeStealEnchantment() {
        super("life_steal", 5, "Life Steal", ItemTags.WEAPON_ENCHANTABLE, EquipmentSlot.MAINHAND);
    }

    @EventHandler
    public void onDamage(EntityDamageByEntityEvent e) {
        if (e.getDamager() instanceof Player damager) {
            ItemStack item = damager.getInventory().getItemInMainHand();
            if (item.containsEnchantment(CustomEnchantments.LIFE_STEAL)) {


                AttributeInstance maxHealth = damager.getAttribute(Attribute.GENERIC_MAX_HEALTH);
                if (maxHealth == null) return;
                double maxHealthValue = maxHealth.getValue();

                double currentHealth = damager.getHealth();

                double heal = e.getDamage() * item.getEnchantmentLevel(CustomEnchantments.LIFE_STEAL) * 0.1;

                damager.setHealth(Math.min(currentHealth + heal, maxHealthValue));
            }
        }
    }
}

File: src/main/kotlin/io/github/chaosdave34/kitpvp/enchantments/impl/LifeStealEnchantment.kt
package io.github.chaosdave34.kitpvp.enchantments.impl

import io.github.chaosdave34.kitpvp.enchantments.CustomEnchantment
import org.bukkit.attribute.Attribute
import org.bukkit.attribute.AttributeInstance
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.entity.EntityDamageByEntityEvent
import org.bukkit.inventory.ItemStack
import kotlin.math.min

class LifeStealEnchantment: CustomEnchantment("life_steal", "Life Steal") {
    @EventHandler
    fun onDamage(event: EntityDamageByEntityEvent) {
        val player = event.damager
        if (player is Player) {
            val item: ItemStack = player.inventory.itemInMainHand
            if (item.containsThisEnchantment()) {
                val maxHealth: AttributeInstance = player.getAttribute(Attribute.GENERIC_MAX_HEALTH) ?: return
                val maxHealthValue = maxHealth.value

                val currentHealth: Double = player.health

                val heal = event.damage * item.getEnchantmentLevel() * 0.1

                player.health = min(currentHealth + heal, maxHealthValue)
            }
        }
    }
}